### PR TITLE
Add support for Inline colors to Gutenberg editor

### PR DESCRIPTION
**Pull request by MuluhGodson**
## Fixes

Fixes https://github.com/creativecommons/project_creativecommons.org/issues/124 by @MuluhGodson 

## Description
Add support for inline colours to the Gutenberg editor and to the Frontend.
This PR accomplishesthe following:
- [x] Move frontend CSS generators into its own function
- [x] Modify frontend CSS action to use the CSS generating function
- [x] Add action for inline colors to use the CSS  generating function

## Technical details
I moved the code that generated the CSS classes into a function so I can use it for both actions that will enable inline colors for the admin editor and also be able to render these colors in the frontend. 

#### Frontend render
image

#### Admin editor
image

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the *default* branch of the repository (`main` or `master`).
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by MuluhGodson** [comment]
I finally resolved it by using an appropriate action hook:
`admin_head` which is the WP recommended hook for adding inline CSS to the admin pages.

https://developer.wordpress.org/reference/hooks/admin_head/

File: functions.php [hunk 1]
```diff
@@ -225,6 +225,38 @@ function enqueue_scripts() {
...
+	   // format styles
+	   $styles = ":root .has-background { background-color: var(--bgColor); }
+	   :root .has-text-color { color: var(--textColor); } :root .has-inline-color { color: var(--textColor); } ";
+	   foreach( $palette[0] as $name => $value ) {
+		   $slug = $value['slug'];
+		   $color = $value['color'];
+	   
+		   $styles .= ".has-{$slug}-background-color { --bgColor: {$color}; } ";
+		   $styles .= ".has-{$slug}-color { --textColor: {$color}; } ";
+	   }
```
Comment (brylie) [inline functions.php:255]: Double-check the indentation in this code block. If possible, use a code linter or formatted to make indentation issues more clear in your editor.

**Review by brylie — CHANGES_REQUESTED** [review]

File: functions.php [hunk 2]
```diff
@@ -225,6 +225,38 @@ function enqueue_scripts() {
...
+	 * This function below takes the colors we have added to 
+	 * the color pallete of wordpress and generates WordPress compatible class names. 
+	 * The styles generated are then added to the HTML pages as Internal CSS styles. 
+	 * We do this by using the Wordpress action (wp_head) which lets us run this function when the head is called.
+	 * @var editor-color-pallete
+	 */
+	function custom_css()
+	{
+	   $palette = get_theme_support( 'editor-color-palette' );
+	   if( !$palette ) { return; } // If our color pallete has not been define don't run this function.
```
Comment (brylie) [inline functions.php:244]:
```suggestion
	 * the color palette of wordpress and generates WordPress compatible class names. 
	 * The styles generated are then added to the HTML pages as Internal CSS styles. 
	 * We do this by using the Wordpress action (wp_head) which lets us run this function when the head is called.
	 * @var editor-color-palette
	 */
	function custom_css()
	{
	   $palette = get_theme_support( 'editor-color-palette' );
	   if( !$palette ) { return; } // If our color palette has not been define don't run this function.
```

**Review by brylie — CHANGES_REQUESTED** [review]

**Comment by brylie** [comment]
Looks good overall.  Just a couple of minor linting issues.

In general, though, I am wondering whether this type of code should run at runtime or would be better to cache somehow. If it turns into a performance issue, we can pre-compute the CSS classes or look into some form of caching.

**Comment by MuluhGodson** [comment]
> In general, though, I am wondering whether this type of code should run at runtime or would be better to cache somehow. If it turns into a performance issue, we can pre-compute the CSS classes or look into some form of caching.

This function runs each time the hook `admin_head` is called. 
If we precompute the CSS classes and put them in the `style.css` file, they will then be loaded each time `style.css` is loaded and this file I believe is cached by WordPress. so that is one way to go, but the downside is that the `style.css` file is loaded only on the client pages. So in effect, if you added an inline colour in the Gutenberg editor, you won't be able to see it but when you render the page and view it on the client-side, you will be able to see the colours. This is one of the reasons why I decided to autogenerate it with the `functions.php` file and have them inserted to the various pages each time their respective hooks are called.

**Comment by MuluhGodson** [comment]
@brylie for performance issues, I have a suggestion.
How likely are the CC colors (as defined on Figma) to change?
If this is unlikely, then we can alternatively remove generating these CSS classes with PHP to avoid performance issues and rather just copy the CSS classes to the `style.css`. 
This file is already loaded on the client side. So we just add it as a stylesheet for the admin side and that should be able to handle the performance issues and we don't have to worry about caching since WP automatically caches the stylesheets.

Then in the documentation, include that if a color is changed or added, the classes in the `style.css` be updated as needed.

**Comment by brylie** [comment]
> How likely are the CC colors (as defined on Figma) to change?

I believe the CC brand color choices are settled for now and wouldn't likely change until another redesign project. @annatuma, are the current brand color choices stable for the duration of this project?

**Comment by MuluhGodson** [comment]
@brylie that's great. I'll do a new commit with the classes in the `style.css` file if that's okay with you. 
I already did locally and tested.

**Comment by MuluhGodson** [comment]
@brylie I just pushed the commit. 
I don't know if the indentation is correct. I am using a CSS formatter for vs-code and while it looks great on vs code, here on GitHub it looks like there's too much indentation. I can manually reduce the indentation.

**Review by brylie — APPROVED** [review]

**Comment by annatuma** [comment]
> > How likely are the CC colors (as defined on Figma) to change?
> 
> I believe the CC brand color choices are settled for now and wouldn't likely change until another redesign project. @annatuma, are the current brand color choices stable for the duration of this project?

Yes, stable for the duration of the project.

**Pull request merged**